File: app/src/main/java/com/example/qiweili/healthapp/Food/Meal.java
package com.example.qiweili.healthapp.Food;

import java.util.ArrayList;

/**
 * Class for meal object. Superclass of Food.java
 * Created by Lucas on 2/13/2018.
 */

public class Meal {

    //array list of foods that makes up the current meal
    private ArrayList<Food> foodList;
    private String mealName = "";
    private String description = "";

    //default constructor method
    public Meal(String mealName) {
        this.mealName = mealName;
        foodList = new ArrayList<>();
    }

    //constructor with a list of foods already given
    public Meal(String mealName, ArrayList<Food> foodList) {
        this.mealName = mealName;
        this.foodList = foodList;
        makeDescription();
    }

    public void addFood(String name, int calories, int protein, int fats) {
        Food food = new Food(name, calories, protein, fats);
        foodList.add(food);
        makeDescription();
    }

    public void deleteFood(Food food) {
        foodList.remove(food);
        makeDescription();
    }

    public ArrayList<Food> getFoodList() {
        return foodList;
    }

    public void setFoodList(ArrayList<Food> foodList) {
        this.foodList = foodList;
        makeDescription();
    }

    //create a list of foods in the current meal
    private void makeDescription() {
        for (int i = 0; i < foodList.size() - 1; i++) {
            description = description + foodList.get(i).getName() + ", ";
        }
        description = description + foodList.get(foodList.size() - 1).getName();
    }

    public String getDescription() {
        return description;
    }

    public void clearDescription() {
        description = "";
    }

    public String getMealName() {
        return mealName;
    }

    public void setMealName(String mealName) {
        this.mealName = mealName;
    }
}

File: app/src/main/java/com/example/qiweili/healthapp/Food/Meal.kt
package com.example.qiweili.healthapp.Food

import java.util.ArrayList

/**
 * Class for meal object. Superclass of Food.java
 * Created by Lucas on 2/13/2018.
 */

class Meal {

    //array list of foods that makes up the current meal
    private var foodList: MutableList<Food>? = null
    val foodMap : MutableMap<String,MutableList<Food>>? = null
    var mealName : String? = null
    var description : String? = null
    var timeStamp : String? = null

    //default constructor method
    constructor(time : String, mealName: String) {
        this.mealName = mealName
        this.timeStamp = time
        foodList = mutableListOf()
        foodMap?.put(timeStamp!!,foodList!!)
    }

    //constructor with a list of foods already given
    constructor(mealName: String, foodList: MutableList<Food>, time : String) {
        this.mealName = mealName
        this.foodList = foodList
        this.timeStamp = time
        foodMap?.put(timeStamp!!,foodList)
        makeDescription()
    }

    fun addFood(name: String, calories: Int, protein: Int, fats: Int) {
        val food = Food(name, calories, protein, fats)
        foodList!!.add(food)
        foodMap?.set(timeStamp!!,foodList!!)
        makeDescription()
    }

    fun deleteFood(food: Food) {
        foodList!!.remove(food)
        foodMap?.set(timeStamp!!,foodList!!)
        makeDescription()
    }

    fun getFoodList(): MutableList<Food>? {
        return foodList
    }

    fun setFoodList(foodList: ArrayList<Food>) {
        this.foodList = foodList
        makeDescription()
    }

    fun getDailyCal(time : String): Int{
        val todaysFood = foodMap?.get(timeStamp)
        var totalCal = 0
        for(i in todaysFood!!.iterator()){
            totalCal += i.calories
        }
        return totalCal
    }

    fun getCalIntake() : MutableList<Int>{
        val food = foodMap?.values
        val time = foodMap?.keys?.iterator()
        val calories = mutableListOf<Int>()
        for(i in food!!) {
            calories.add(getDailyCal(time?.next()!!))

        }
        return calories
    }

    //create a list of foods in the current meal
    private fun makeDescription() {
        for (i in 0 until foodList!!.size - 1) {
            description = description + foodList!![i].name + ", "
        }
        description = description + foodList!![foodList!!.size - 1].name
    }

    fun clearDescription() {
        description = ""
    }
}